src/stan/services/optimize/bfgs.hpp

```diff
@@ -0,0 +1,202 @@
+        bfgs._ls_opts.alpha0 = init_alpha;
+        bfgs._conv_opts.tolAbsF = tol_obj;
+        bfgs._conv_opts.tolRelF = tol_rel_obj;
+        bfgs._conv_opts.tolAbsGrad = tol_grad;
+        bfgs._conv_opts.tolRelGrad = tol_rel_grad;
+        bfgs._conv_opts.tolAbsX = tol_param;
+        bfgs._conv_opts.maxIts = num_iterations;
+
+        double lp = bfgs.logp();
+
+        std::stringstream msg;
+        msg << "Initial log joint probability = " << lp;
+        message_writer(msg.str());
+
+        std::vector<std::string> names;
+        names.push_back("lp__");
+        model.constrained_param_names(names, true, true);
+        parameter_writer(names);
+
+        if (save_iterations) {
+          std::vector<double> values;
+          msg.str("");
+          model.write_array(rng, cont_vector, disc_vector, values,
+                            true, true, &msg);
+          if (msg.str().length() > 0)
+            message_writer(msg.str());
+
+          values.insert(values.begin(), lp);
+          parameter_writer(values);
+        }
+        int ret = 0;
+
+        while (ret == 0) {
+          interrupt();
+          if (refresh > 0
+              && (bfgs.iter_num() == 0
+                  || (bfgs.iter_num() + 1 % refresh == 0)))
+            message_writer("    Iter "
+                           "     log prob "
+                           "       ||dx|| "
+                           "     ||grad|| "
+                           "      alpha "
+                           "     alpha0 "
+                           " # evals "
+                           " Notes ");
+
+          ret = bfgs.step();
+          lp = bfgs.logp();
+          bfgs.params_r(cont_vector);
+
+          if (refresh > 0
+              && (ret != 0
+                  || !bfgs.note().empty()
+                  || bfgs.iter_num() == 0
+                  || (bfgs.iter_num() + 1 % refresh == 0))) {
+            msg.str("");
+            msg << " " << std::setw(7) << bfgs.iter_num() << " ";
+            msg << " " << std::setw(12) << std::setprecision(6)
+                << lp << " ";
+            msg << " " << std::setw(12) << std::setprecision(6)
+                << bfgs.prev_step_size() << " ";
+            msg << " " << std::setw(12) << std::setprecision(6)
+                << bfgs.curr_g().norm() << " ";
+            msg << " " << std::setw(10) << std::setprecision(4)
+                << bfgs.alpha() << " ";
+            msg << " " << std::setw(10) << std::setprecision(4)
+                << bfgs.alpha0() << " ";
+            msg << " " << std::setw(7)
+                << bfgs.grad_evals() << " ";
+            msg << " " << bfgs.note() << " ";
+            message_writer(msg.str());
+            msg.str("");
+          }
+
+          if (bfgs_ss.str().length() > 0) {
+            message_writer(bfgs_ss.str());
+            bfgs_ss.str("");
+          }
+
+          if (save_iterations) {
+            std::vector<double> values;
+            msg.str("");
```

Comment:
We can fix this with using local streams. I'll see how hard it is to switch things over.